File: backend/src/matching/pipeline/money-field-bindings.ts
/**
 * Feature 010 — which questionnaire answer feeds which economic input.
 *
 * This is a CODE CONSTANT, deliberately not a column on `Question`.
 * Anti-Pattern A33 forbids scoring / eligibility / profile-mapping fields on
 * `Question` / `QuestionOption`, and Constitution v6.0.0 specifically deleted
 * `Question.profileField` / `Question.systemRole`. Storing the binding would
 * revive exactly that. Renaming a bound question code is therefore a code
 * change — the accepted residual limit recorded in the spec.
 *
 * The four NUMERIC questions named here replace the old bucket→midpoint maps
 * (a customer asking for 500 000 was quoted on 300 000). A bound answer that is
 * absent is NEVER substituted with a default: the quote yields no figures
 * (`MONEY_FIGURE_MISSING`, FR-044).
 *
 * `existing_obligations` is special: its answer is DERIVED, not typed. See
 * `resolveObligations` below — the applicant states each debt separately and the
 * total is summed, because one lump sum recalled from memory is the least
 * reliable input in the flow and the one with the largest effect on DBR.
 *
 * Mirrored on mobile by the four `*_apply_mapper.dart` files.
 */
import { Decimal } from '@prisma/client/runtime/library';

export const MONEY_FIELD_BINDINGS = {
  requested_amount: 'amount_requested',
  tenor_months: 'repayment_period_months',
  monthly_income: 'monthly_income',
  existing_obligations: 'current_installments',
} as const;

export type MoneyFieldBinding = keyof typeof MONEY_FIELD_BINDINGS;

/** The question code each binding expects, for publish-time validation. */
export type BoundQuestionCode = (typeof MONEY_FIELD_BINDINGS)[MoneyFieldBinding];

export const MONEY_FIELD_BINDING_KEYS = Object.keys(
  MONEY_FIELD_BINDINGS,
) as readonly MoneyFieldBinding[];

/**
 * The engine field each binding lands on, and the bounds the seed applies.
 * `path` is documentation for reviewers (and the `meta.binding` audit trail) —
 * it is not evaluated dynamically.
 */
export const MONEY_FIELD_BINDING_SPECS: Readonly<
  Record<
    MoneyFieldBinding,
    {
      readonly questionCode: BoundQuestionCode;
      readonly path: string;
      readonly unit: 'EGP' | 'months';
      readonly integerOnly: boolean;
    }
  >
> = Object.freeze({
  requested_amount: {
    questionCode: MONEY_FIELD_BINDINGS.requested_amount,
    path: 'requestedAmountEGP',
    unit: 'EGP',
    integerOnly: false,
  },
  tenor_months: {
    questionCode: MONEY_FIELD_BINDINGS.tenor_months,
    path: 'preferredTenorMonths',
    unit: 'months',
    integerOnly: true,
  },
  monthly_income: {
    questionCode: MONEY_FIELD_BINDINGS.monthly_income,
    path: 'employment.monthlyNetSalaryEGP',
    unit: 'EGP',
    integerOnly: false,
  },
  existing_obligations: {
    questionCode: MONEY_FIELD_BINDINGS.existing_obligations,
    path: 'obligations.existingMonthlyObligationsEGP',
    unit: 'EGP',
    integerOnly: false,
  },
});

// ---- Itemised obligations ---------------------------------------------------
//
// The applicant picks WHICH debts he carries in one MULTI_SELECT, and each picked
// option unlocks its own NUMERIC amount question through the questionnaire's
// existing `enabledWhen` branching (`question-visibility.ts` already matches an
// option code against a multi-pick answer). So this needs no new question type,
// no new table and no schema change — only the mapping below, which is the
// single source of truth shared by the seed, the sum, and publish validation.
//
// Deliberately code constants, not columns: A33 forbids profile-mapping fields on
// `Question`, and the four bindings above are already code for the same reason.

/**
 * The MULTI_SELECT whose picks decide which amount questions are asked.
 *
 * This is the pool's PRE-EXISTING "Do you pay back any loans right now?" — not a
 * new question. It already asks exactly this, with exactly these option codes, so
 * introducing a second debt-type multi-select would ask the applicant the same
 * thing twice. The option codes below are therefore that question's codes
 * verbatim (note `credit_cards`, plural, as authored).
 */
export const DEBT_TYPES_QUESTION_CODE = 'current_loans';

/**
 * The explicit "I have none" pick. Present so "no debts" is a STATED answer
 * rather than an empty one — an empty multi-select is indistinguishable from an
 * unanswered question, and that difference decides whether obligations resolve
 * to a real `0` or to "no figures".
 */
export const DEBT_TYPE_NONE_OPTION = 'none';

/** Debt-type option code → the NUMERIC question that captures its instalment. */
export const OBLIGATION_ITEM_QUESTION_BY_DEBT_TYPE = Object.freeze({
  car_loan: 'obligation_car_loan',
  credit_cards: 'obligation_credit_card',
  personal_loan: 'obligation_personal_loan',
  mortgage: 'obligation_mortgage',
  other: 'obligation_other',
} as const);

export type DebtTypeOptionCode = keyof typeof OBLIGATION_ITEM_QUESTION_BY_DEBT_TYPE;
export type ObligationItemQuestionCode =
  (typeof OBLIGATION_ITEM_QUESTION_BY_DEBT_TYPE)[DebtTypeOptionCode];

export const DEBT_TYPE_OPTION_CODES = Object.keys(
  OBLIGATION_ITEM_QUESTION_BY_DEBT_TYPE,
) as readonly DebtTypeOptionCode[];

export const OBLIGATION_ITEM_QUESTION_CODES = Object.values(
  OBLIGATION_ITEM_QUESTION_BY_DEBT_TYPE,
) as readonly ObligationItemQuestionCode[];

/** The itemised amount question a debt-type pick unlocks, if the pick is known. */
export function obligationItemQuestionFor(
  debtTypeOptionCode: string,
): ObligationItemQuestionCode | undefined {
  return debtTypeOptionCode in OBLIGATION_ITEM_QUESTION_BY_DEBT_TYPE
    ? OBLIGATION_ITEM_QUESTION_BY_DEBT_TYPE[debtTypeOptionCode as DebtTypeOptionCode]
    : undefined;
}

export interface ObligationsResolution {
  /** Authoritative monthly obligations — what DBR is computed on. */
  totalEGP: Decimal;
  /** Whether the total came from itemised answers rather than a stated lump sum. */
  itemised: boolean;
  /** The item questions that carried a value, biggest-first ordering not implied. */
  itemisedCodes: readonly string[];
  /**
   * Derived from the debt-type picks when itemised, else from `total > 0`. A debt
   * whose instalment is currently 0 still means a loan is on book, which the
   * lump-sum derivation could never express.
   */
  hasCurrentLoan: boolean;
  /**
   * Set when the client also submitted `current_installments` and it disagrees
   * with the itemised sum. Returned rather than thrown: this module is pure, the
   * same way `quoteProgram` reports `unavailable` instead of raising. Callers
   * raise `OBLIGATIONS_TOTAL_MISMATCH`.
   */
  statedTotalMismatch: { statedEGP: Decimal; computedEGP: Decimal } | null;
}

/** Sums to the cent, so a mismatch below this is float noise, not disagreement. */
const TOTAL_TOLERANCE = new Decimal('0.01');

/**
 * Resolve the applicant's monthly obligations from questionnaire answers.
 *
 * Itemised path (the debt-type question was asked and answered): the total is the
 * SUM of the visible per-debt answers. It is authoritative — a client-submitted
 * `current_installments` is only cross-checked against it, never preferred, so
 * editing the total cannot buy affordability.
 *
 * Fallback path (the snapshot serves no debt-type question, i.e. it predates this
 * feature, or a historical application's answers): read `current_installments`
 * directly, exactly as before. An un-republished questionnaire keeps working.
 *
 * Returns `null` when neither path yields a figure — callers must then produce no
 * figures rather than substitute a default (FR-044).
 */
export function resolveObligations(args: {
  /** Every NUMERIC answer by question code, already validated. */
  readonly numericByCode: ReadonlyMap<string, string>;
  /** The debt-type picks, or `undefined` when that question was not asked. */
  readonly pickedDebtTypes?: readonly string[];
}): ObligationsResolution | null {
  const { numericByCode, pickedDebtTypes } = args;
  const statedRaw = numericByCode.get(MONEY_FIELD_BINDINGS.existing_obligations);
  const stated = statedRaw !== undefined ? new Decimal(statedRaw) : null;

  if (pickedDebtTypes === undefined) {
    if (stated === null) return null;
    return {
      totalEGP: stated,
      itemised: false,
      itemisedCodes: [],
      hasCurrentLoan: stated.greaterThan(0),
      statedTotalMismatch: null,
    };
  }

  // Only picks that map to an amount question can contribute. `none` maps to
  // nothing by construction, so picking it sums to a real, stated zero.
  const itemisedCodes: string[] = [];
  let total = new Decimal(0);
  for (const pick of pickedDebtTypes) {
    const itemCode = obligationItemQuestionFor(pick);
    if (itemCode === undefined) continue;
    const raw = numericByCode.get(itemCode);
    if (raw === undefined) continue;
    itemisedCodes.push(itemCode);
    total = total.plus(new Decimal(raw));
  }

  const carriesDebt = pickedDebtTypes.some(
    (pick) => pick !== DEBT_TYPE_NONE_OPTION && obligationItemQuestionFor(pick) !== undefined,
  );

  return {
    totalEGP: total,
    itemised: true,
    itemisedCodes,
    hasCurrentLoan: carriesDebt,
    statedTotalMismatch:
      stated !== null && stated.minus(total).abs().greaterThan(TOTAL_TOLERANCE)
        ? { statedEGP: stated, computedEGP: total }
        : null,
  };
}

File: masrafy-app/lib/features/questionnaire/domain/constants/money_field_bindings.dart
/// Which questionnaire answer feeds which economic input — the mobile mirror of
/// `backend/src/matching/pipeline/money-field-bindings.ts`.
///
/// These are CODE CONSTANTS on both sides on purpose: Anti-Pattern A33 forbids
/// scoring / eligibility / profile-mapping fields on `Question`, so the binding
/// can never live on the question row. Renaming a bound question code is
/// therefore a code change on both platforms — the accepted residual limit
/// recorded in the feature-010 spec.
///
/// Each bound question is NUMERIC, so the applicant states the real figure. The
/// bucket→midpoint maps these replaced quoted a customer asking for 500 000 on
/// 300 000; a missing bound answer is NEVER substituted with a default (FR-044)
/// — the flow blocks instead.
library;

/// `requestedAmountEGP` — the principal the applicant asked for.
const String kRequestedAmountQuestion = 'amount_requested';

/// `preferredTenorMonths` — repayment period, already expressed in months.
const String kTenorMonthsQuestion = 'repayment_period_months';

/// `employment.monthlyNetSalaryEGP` — monthly income.
const String kMonthlyIncomeQuestion = 'monthly_income';

/// `obligations.existingMonthlyObligationsEGP` — the applicant's total monthly
/// instalments.
///
/// DERIVED, not typed: it is the sum of [kObligationItemQuestionByDebtType], so
/// the field renders read-only and the applicant never recalls a lump sum. The
/// backend re-sums it and rejects a disagreement (`OBLIGATIONS_TOTAL_MISMATCH`),
/// so editing it in flight cannot buy affordability.
const String kExistingObligationsQuestion = 'current_installments';

/// Every bound question code, for completeness checks before submitting.
const List<String> kMoneyFieldQuestionCodes = [
  kRequestedAmountQuestion,
  kTenorMonthsQuestion,
  kMonthlyIncomeQuestion,
  kExistingObligationsQuestion,
];

// ---- Itemised obligations ---------------------------------------------------
//
// The applicant ticks WHICH debts he carries in one MULTI_SELECT, and each pick
// unlocks its own amount question through the snapshot's existing `enabledWhen`
// branching — so this needs no new question type and no new widget. The mapping
// below mirrors `OBLIGATION_ITEM_QUESTION_BY_DEBT_TYPE` on the backend, which is
// also what the seed and publish validation read.

/// The MULTI_SELECT whose picks decide which amount questions are asked.
///
/// The pool's PRE-EXISTING "Do you pay back any loans right now?" — not a new
/// question. Option codes are that question's own (note `credit_cards`, plural).
const String kDebtTypesQuestion = 'current_loans';

/// The explicit "I have none" pick, so no-debt is a STATED answer. An empty
/// multi-select is indistinguishable from an untouched one, and that difference
/// decides whether obligations resolve to a real 0 or to "no figures".
const String kDebtTypeNoneOption = 'none';

/// The pick whose amount question states a LIMIT rather than an instalment.
const String kCreditCardDebtTypeOption = 'credit_cards';

/// Total credit limit across EVERY card the applicant holds, at every bank.
///
/// A card has no fixed monthly payment, and an undrawn limit is money that can be
/// drawn tomorrow, so the minimum payment on today's statement understates what a
/// bank underwrites against. The limit is stated instead and discounted by
/// [kCreditCardLimitMonthlyPercent].
const String kCreditCardLimitQuestion = 'credit_card_total_limit';

/// Share of the stated card limit that counts as a monthly commitment. Mirrors
/// `CREDIT_CARD_LIMIT_MONTHLY_PERCENT` on the backend — the server re-derives the
/// same total and rejects a disagreement (`OBLIGATIONS_TOTAL_MISMATCH`), so the
/// two constants must move together.
const double kCreditCardLimitMonthlyPercent = 5;

/// Debt-type option code → the NUMERIC question capturing its monthly figure.
const Map<String, String> kObligationItemQuestionByDebtType = {
  'car_loan': 'obligation_car_loan',
  kCreditCardDebtTypeOption: kCreditCardLimitQuestion,
  'personal_loan': 'obligation_personal_loan',
  'mortgage': 'obligation_mortgage',
  'other': 'obligation_other',
};

/// Every per-debt amount question code.
final List<String> kObligationItemQuestionCodes =
    kObligationItemQuestionByDebtType.values.toList(growable: false);

/// The amount question a debt-type pick unlocks, or `null` for `none`/unknown.
String? obligationItemQuestionFor(String debtTypeOptionCode) =>
    kObligationItemQuestionByDebtType[debtTypeOptionCode];

/// The monthly commitment one stated per-debt figure contributes.
///
/// Identity for every debt whose answer already IS a monthly instalment; for
/// credit cards the stated figure is a limit, so it is discounted (150,000 of
/// limit → 7,500 a month). An unmapped pick passes through at face value rather
/// than dropping out of the total.
///
/// Rounded to two decimals so it lands on the server's `Decimal(18, 2)` column
/// and matches the `ROUND_HALF_UP` the backend applies to the same product.
double obligationMonthlyAmountFor(String debtTypeOptionCode, double statedEGP) {
  if (debtTypeOptionCode != kCreditCardDebtTypeOption) return statedEGP;
  return double.parse(
    (statedEGP * kCreditCardLimitMonthlyPercent / 100).toStringAsFixed(2),
  );
}
